client/state/routing/middleware.js

```diff
@@ -7,50 +7,47 @@ import page from 'page';
 /**
  * Internal dependencies
  */
-import localforage from 'lib/localforage';
-import { isOutsideCalypso } from 'lib/url';
+import {
+	getSavedPath,
+	savePath,
+} from 'lib/restore-last-path';
 import { ROUTE_SET } from 'state/action-types';
 
 const debug = debugFactory( 'calypso:restore-last-location' );
-const LAST_PATH = 'last_path';
-const ALLOWED_PATHS_FOR_RESTORING = /^\/(stats|plans|view|posts|pages|media|types|themes|sharing|people|plugins|domains)/i;
-
-const isWhitelistedForRestoring = ( path ) => {
-	return !! path.match( ALLOWED_PATHS_FOR_RESTORING );
-};
-
-export const restoreLastLocation = () => {
-	let hasInitialized = false;
+let hasInitialized = false;
 
+export const routingMiddleware = () => {
 	return ( next ) => ( action ) => {
 		if ( action.type !== ROUTE_SET || ! action.path || ! action.query ) {
 			return next( action );
 		}
 
-		localforage.getItem( LAST_PATH ).then(
-			( lastPath ) => {
-				if ( ! hasInitialized &&
-						lastPath && lastPath !== '/' &&
-						action.path === '/' && Object.keys( action.query ).length === 0 &&
-						! isOutsideCalypso( lastPath ) &&
-						isWhitelistedForRestoring( lastPath ) ) {
-					debug( 'redir to', lastPath );
-					page( lastPath );
-				} else if ( action.path !== lastPath &&
-						! isOutsideCalypso( action.path ) ) {
-					debug( 'saving', action.path );
-					localforage.setItem( LAST_PATH, action.path );
-				}
-
-				if ( ! hasInitialized ) {
-					hasInitialized = true;
-				}
-
-				return next( action );
-			},
-			() => next( action )
-		);
+		if ( Object.keys( action.query ).length !== 0 ) {
+			return next( action );
+		}
+
+		const firstRun = ! hasInitialized;
```

Comment:
Tiny nit - might this be ever so slightly clearer as `isFirstRun`?
FFTI this though.